app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/OffboardRouterActivityJava.java

```diff
@@ -0,0 +1,225 @@
+      destination = Point.fromLngLat(point.getLongitude(), point.getLatitude());
+      mapboxMap.addMarker(new MarkerOptions().position(point));
+      findRoute();
+    } else if (waypoint == null) {
+      waypoint = Point.fromLngLat(point.getLongitude(), point.getLatitude());
+      mapboxMap.addMarker(new MarkerOptions().position(point));
+      findRoute();
+    } else {
+      Toast.makeText(this, "Only 2 waypoints supported", Toast.LENGTH_LONG).show();
+      clearMap();
+    }
+    return false;
+  }
+
+  private void clearMap() {
+    if (mapboxMap != null) {
+      mapboxMap.clear();
+      route = null;
+      destination = null;
+      waypoint = null;
+      navigationMapRoute.updateRouteVisibilityTo(false);
+      navigationMapRoute.updateRouteArrowVisibilityTo(false);
+    }
+  }
+
+  private void findRoute() {
+    Router offboardRouter = new MapboxOffboardRouter(
+      NavigationRoute.builder(this)
+    );
+    directionsSession = new MapboxDirectionsSession(
+      offboardRouter,
+      this
+    );
+    if (origin != null && destination != null) {
+      if (TurfMeasurement.distance(origin, destination, TurfConstants.UNIT_METERS) > 50) {
+        List<Point> waypoints = new ArrayList<>();
+        if (waypoint != null) {
+          waypoints.add(waypoint);
+        }
+        RouteOptionsNavigation.Builder optionsBuilder = new RouteOptionsNavigation.Builder();
+        optionsBuilder.accessToken((Utils.getMapboxAccessToken(this)));
+        optionsBuilder.origin(origin);
+        optionsBuilder.destination(destination);
+        for (Point waypointPoint : waypoints) {
+          optionsBuilder.addWaypoint(waypointPoint);
+        }
+        directionsSession.requestRoutes(optionsBuilder.build());
+      }
+    }
+  }
+
+  /*
+   * DirectionSessions.RouteObserver
+   */
+
+  @Override
+  public void onRoutesChanged(@NotNull List<Route> routes) {
+    if (!routes.isEmpty()) {
+      route = Mappers.mapToDirectionsRoute(routes.get(0));
+      navigationMapRoute.addRoute(route);
+    }
+  }
+
+  @Override
+  public void onRoutesRequested() {
+    Timber.d("onRoutesRequested: navigation.getRoute()");
+  }
+
+  @Override
+  public void onRoutesRequestFailure(@NotNull Throwable throwable) {
+    Timber.e(throwable, "onRoutesRequestFailure: navigation.getRoute()");
+  }
+
+  /*
+   * Activity lifecycle methods
+   */
+
+  @Override
+  public void onResume() {
+    super.onResume();
+    mapView.onResume();
+  }
+
+  @Override
+  public void onPause() {
+    super.onPause();
+    mapView.onPause();
+  }
+
+  @Override
+  protected void onStart() {
+    super.onStart();
+    mapView.onStart();
+  }
+
+  @Override
+  protected void onStop() {
+    super.onStop();
+    mapView.onStop();
+  }
+
+  @Override
+  public void onLowMemory() {
+    super.onLowMemory();
+    mapView.onLowMemory();
+  }
+
+  @Override
+  protected void onDestroy() {
+    super.onDestroy();
+    if (directionsSession != null) {
+      directionsSession.cancel();
```

Comment:
Now that I'm looking at this I realized that we're exposing `cancel` (and `getRoute`) from both `Router` and `DirectionsSession` and `DirectionsSession` is only delegating to `Router`. Don't you think that this could be confusing to clients as they'd need to create both the `Router` and the `DirectionsSession` (like we did in this example) cause both functions are doing the same? What happens if both are called (I believe nothing) but we should think if this could be restricted somehow to avoid misusing the APIs exposed. I know that we could add documentation but in general we should design APIs that are hard to misuse. What do you think? Any ideas?

Reply:
Yeah, I'm agree, it can confuse users... As option, we can pass `LifecycleOwner` to `MapboxDirectionsSession` constructor, and then:
1) We will care by ourself about erasing all stuff we need (router for example)
2) We can remove `MapboxDirectionsSession#cancel` method, and leave only `MapboxOffboardRouter#cancel` method to save possibility to users cancel the router call

wdyt?

Reply:
Going to write this concern down and we can revisit later so we don't block outstanding `Offboard` PRs. But yeah, we definitely need to discuss this before `1.0` goes out.